### ISSUE TITLE
New method: calculate all possible distances, angles, dihedrals

### ISSUE THREAD

**Issue by berquist** [cclib/cclib#489]
As discussed on the dev mailing list, we should have a method that calculates all possible distances, angles, dihedrals, etc.

Possibly useful references:
* https://github.com/CrawfordGroup/ProgrammingProjects/tree/master/Project%2301
* https://github.com/berquist/programming_party/blob/master/eric/molecule.py
* https://stackoverflow.com/q/20305272/3249688

From Karol:
> Regarding the coordinate-to-angle calculations - they'll be so fast for 99% of the cases that it's not worthwhile to think about calculating things on-the-fly.

So things should be calculated all at once.

**Comment by LaTruelle** [comment]
I used it also on a project of mine, see:

https://github.com/chemistry-scripts/computations-scripts/blob/master/ASM_NBO_Gaussian.py#L161-L200

It uses numpy in the process, but it is already used by cclib so should not be a problem.

Main issue will be to find an elegant/efficient way to store the data. I actually prefer using simple methods so computing on the fly is best in that case, but that should not be chosen for speed purposes indeed, more for ease of use.

**Comment by fdroessler** [comment]
Is this being worked on by someone? Otherwise, I'd be happy to add a PR as I have been working on something similar anyway.

**Comment by Kaiwalya2502** [comment]
Is anyone working on this?
If not, I just have to implement the methods to calculate the bond lengths, angles, and dihedrals from the data given in an input file right?

**Comment by langner** [comment]
> Is anyone working on this?
> If not, I just have to implement the methods to calculate the bond lengths, angles, and dihedrals from the data given in an input file right?

Correct, probably within one of the existing methods modules: https://github.com/cclib/cclib/tree/master/cclib/method

**Comment by GomaruSai** [comment]
Hi!
I would like to know if anyone is working on this issue as I see no PR  yet?
@Kaiwalya2502 Are you working on this?

**Comment by GomaruSai** [comment]
> As discussed on the dev mailing list, we should have a method that calculates **all possible** distances, angles, dihedrals, etc.

Do you mean only bond parameters or "all possible" parameters, Sir? @berquist

**Comment by langner** [comment]
> The interface should be via class methods that appear to use 2, 3, and 4 indices, so that we can start with storing things in 2, 3 and 4-index arrays, but change later if need be. Storing a 4-index array for 100 atoms is not going to be a good approach.

@berquist Can you elaborate what you had in mind here?

**Comment by berquist** [comment]
Something like methods
```python
def bond(self, i: int, j: int) -> float:
    ...
def angle(self, i: int, j: int, k: int) -> float:
    ...
```
for a user interface. Ignore the stuff I said about the implementation until we actually see a working implementation.
